File: backend/app/schemas/assets.py
"""Asset schemas."""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict


class AssetBase(BaseModel):
    asset_tag: str
    name: str
    category_id: int
    serial_number: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    purchase_date: Optional[date] = None
    purchase_cost: Optional[Decimal] = None
    current_location_id: Optional[int] = None
    department_id: Optional[int] = None
    status: str = "active"
    warranty_start: Optional[date] = None
    warranty_end: Optional[date] = None
    amc_start: Optional[date] = None
    amc_end: Optional[date] = None
    vendor_id: Optional[int] = None
    criticality: Optional[str] = None


class AssetCreate(AssetBase):
    pass


class AssetUpdate(BaseModel):
    asset_tag: Optional[str] = None
    name: Optional[str] = None
    category_id: Optional[int] = None
    serial_number: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    purchase_date: Optional[date] = None
    purchase_cost: Optional[Decimal] = None
    current_location_id: Optional[int] = None
    department_id: Optional[int] = None
    status: Optional[str] = None
    warranty_start: Optional[date] = None
    warranty_end: Optional[date] = None
    amc_start: Optional[date] = None
    amc_end: Optional[date] = None
    vendor_id: Optional[int] = None
    criticality: Optional[str] = None


class AssetResponse(AssetBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    created_at: datetime
    updated_at: datetime


class AssetListResponse(BaseModel):
    """Minimal asset for list views."""

    model_config = ConfigDict(from_attributes=True)
    id: int
    asset_tag: str
    name: str
    category_id: int
    status: str
    current_location_id: Optional[int] = None

File: frontend/src/api/assets.ts
import { apiGet } from './client'

export interface Asset {
  id: number
  asset_tag: string
  name: string
  category_id: number
  serial_number: string | null
  manufacturer: string | null
  model: string | null
  purchase_date: string | null
  purchase_cost: string | null
  current_location_id: number | null
  department_id: number | null
  status: string
  warranty_start: string | null
  warranty_end: string | null
  amc_start: string | null
  amc_end: string | null
  vendor_id: number | null
  criticality: string | null
  created_at: string
  updated_at: string
}

export interface Paginated<T> {
  items: T[]
  total: number
  page: number
  size: number
  pages: number
}

export function getAssets(params?: { page?: number; size?: number; search?: string; status?: string }) {
  const searchParams: Record<string, string> = {}
  if (params?.page) searchParams.page = String(params.page)
  if (params?.size) searchParams.size = String(params.size)
  if (params?.search) searchParams.search = params.search
  if (params?.status) searchParams.status = params.status
  return apiGet<Paginated<Asset>>('/assets', Object.keys(searchParams).length ? searchParams : undefined)
}

export function getAsset(id: number) {
  return apiGet<Asset>(`/assets/${id}`)
}
